### DISCUSSION TITLE
showtext Font Support & Emoji Rendering

### DISCUSSION THREAD

**Discussion by rsh52** [ggplot2-extenders/ggplot-extension-club#60] [Q&A] [answered]
Hello friends!

I'm working on expanding {ggswim} to include support for FontAwesome icons. I modelled the set up for this after the {[emojifont](https://github.com/GuangchuangYu/emojifont/tree/master)} package, allowing the .TTF file data to be available on load of the package using `font_add()`/`showtext_auto()` from the {showtext} package. Unfortunately, something in {showtext} causes [the AGG rendering for emojis](https://github.com/yixuan/showtext/issues/54) to break.

I'm curious if anyone has a better solution here. Maybe @yjunechoe has insight considering your excellent [blog post](https://yjunechoe.github.io/posts/2021-06-24-setting-up-and-debugging-custom-fonts/), perhaps there's a way to translate the methods here into the package instead of the users system?

I also posted this on [SO]() and will give an example below.

<details>
  <summary>Click Me for Example</summary>

```r
library(ggplot2)
library(dplyr)

df <- cars |>
  mutate(
    glyph = case_when(speed > 16 ~ "✅",
                      .default = "❌ "),
    label = case_when(speed > 16 ~ "Fast",
                      .default = "Slow"),
    color = case_when(speed > 16 ~ "red",
                      .default = "green")
  ) |>
  tibble::tibble()

# Before showtext
ggplot() +
  geom_text(
    data = df,
    aes(x = speed, y = dist, label = glyph, colour = color),
    size = 5, family = "fa"
  )
```

![image](https://github.com/user-attachments/assets/0b8b237c-319d-42f6-b98a-5ef528bbdb7d)

```r
showtext::showtext_auto(enable = TRUE)

# After showtext
ggplot() +
  geom_text(
    data = df,
    aes(x = speed, y = dist, label = glyph, colour = color),
    size = 5, family = "fa"
  )
```
![image](https://github.com/user-attachments/assets/ee98d116-48c8-4517-bfd1-e2f120174a08)

</details>

**Comment by yjunechoe** [comment, marked as answer]
I've done this in the past (albeit kind of hastily, but I hope it's helpful/readable in its current state!) - https://github.com/yjunechoe/penngradlings/blob/ef4e0e3863045c811d7e049eba125a674e058bbd/R/font-load.R#L39-L89

Essentially, you bundle the fonts with your package under `/inst` and then use `systemfonts::register_font()` to add fonts for use with `{ragg}`.

Cara Thompson's the expert in this area - her blog post might help as well: https://www.cararthompson.com/posts/2024-01-12-using-fonts-in-r-for-dataviz/2024-01-12_getting-fonts-to-work

**Reply by rsh52** [reply]
Ah great, let me try this out and get back to you. I actually found your post through Cara's!

**Reply by teunbrand** [reply]
If you bundle a font with your package, make sure to read the copyright info on that font carefully and check if you are allowed to distribute the font.

**Reply by rsh52** [reply]
This worked _**perfectly**_! Deserves highlighting somewhere, I'm astounded by it.

```r
df <- cars |>
  mutate(
    glyph = case_when(speed > 16 ~ "✅",
                      .default = fontawesome("fa-car")),
    label = case_when(speed > 16 ~ "Fast",
                      .default = "Slow"),
    color = case_when(speed > 16 ~ "red",
                      .default = "green")
  ) |>
  tibble::tibble()

ggplot() +
  geom_swim_marker(
    data = df,
    aes(x = speed, y = dist, marker = label),
    size = 5, family = "FontAwesome"
  ) +
  scale_marker_discrete(glyphs = df$glyph,
                        colours = df$color,
                        limits = df$label)
```

![image](https://github.com/user-attachments/assets/f465e856-1c3f-45a6-937b-8a2e2dafceaf)

**Reply by rsh52** [reply]
> If you bundle a font with your package, make sure to read the copyright info on that font carefully and check if you are allowed to distribute the font.

Good point. My understanding is if I include the FontAwesome `LICENSE.txt` file in the package it should be ok for this which for now I've kept under `/inst` with everything else.

**Reply by yjunechoe** [reply]
Oh yes, make sure to bundle the license as well of course (doesn't hurt!). BTW "OFL" (Open Font License) is the MIT of the fonts world and allows redistribution (the free fontawesome fonts are also licensed under OFL if I remember correctly)

**Reply by rsh52** [reply]
@yjunechoe Quick question, what does `features` do at baseline for `font_features()`:

https://github.com/yjunechoe/penngradlings/blob/ef4e0e3863045c811d7e049eba125a674e058bbd/R/font-load.R#L61

I probably will stick to just FA/BS (for now) so don't think I need the rest, but it's not clear to me what "kern" and "zero" do outside of filling the "4 character" criteria under the hood for `font_features()`. Is it a placeholder?

**Reply by yjunechoe** [reply]
I can't tell you off the top of my head - they're features available for some of the fonts I use and I think they just make nice defaults! 🙃 

Everything I know about typography in R I learned from copying Bob Rudis 🤣  https://rud.is/b/2021/02/19/getting-into-the-rhythm-of-chart-typography-with-ragg-and-hrbragg-a-k-a-its-ragg-time/

**Reply by rsh52** [reply]
Ah wonderful, thank you! I don't blame you too much for not knowing everything immediately, speaking as someone who knows very little always. Appreciate the resource.

**Comment by teunbrand** [comment]
Systemfonts now has some extra functionality around required fonts. See https://www.tidyverse.org/blog/2025/01/text-rendering-updates/#working-with-non-installed-fonts

**Reply by rsh52** [reply]
Oh wonderful, thank you for the update! Might be a better way to optimize how the fonts get loaded. I'll add it to a feature consideration.
